### PR TITLE
Fix type identifiers for RationalNumberType (on negative numbers)

### LINKED ISSUE #4706: Crash when converting signed rational using ABIEncoderV2
_This issue is being reported as part of the current audit being held by Zeppelin Solutions to Solidity's compiler (tag v0.4.24)._

## Description 
Compiler crashes on the assembly code generation when using `pragma experimental ABIEncoderv2` to encode a signed rational.

## Error in file
`libsolidity/codegen/CompilerContext.cpp(368)`

## Displayed message

```
Error parsing/analyzing inline assembly block.
Throw in function void dev::solidity::CompilerContext::appendInlineAssembly(const st
ring&, const std::vector<std::__cxx11::basic_string<char> >&, bool)
------------------ Errors: ----------------
--CODEGEN--:9:35: Error: Expected '(' but got '-'
                        function abi_encode_t_rational_-1337_by_1_to_t_int16(value, pos) {    
```

## Example code

```solidity
pragma solidity ^0.4.24;
pragma experimental ABIEncoderV2;
contract CrashContract {
   function f1() public pure returns (bytes) {
       return abi.encode(1,-2);
   }
}
```

### PR DESCRIPTION
**Pull request by axic**
Fixes #4706.

Needs abi encoder tests.

### CODE REVIEW

**Review by chriseth — APPROVED** [review]

**Review by chriseth — CHANGES_REQUESTED** [review]
Perhaps we could add an assertion to `escapeIdentifier` that the identifier is a valid identifier?

**Comment by axic** [comment]
> Perhaps we could add an assertion to escapeIdentifier that the identifier is a valid identifier?

Good idea, but I'd do it in a separate PR. Created #4732

**Comment by chriseth** [comment]
Tests fail.

**Comment by chriseth** [comment]
Packed encoding is not supported for literals.

**Comment by axic** [comment]
Fixed.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -12527,6 +12527,44 @@ BOOST_AUTO_TEST_CASE(abi_encode_empty_string_v2)
...
+}
+
+BOOST_AUTO_TEST_CASE(abi_encode_rational_v2)
+{
+	char const* sourceCode = R"(
+		// Tests that rational numbers (even negative ones) are encoded properly.
+		pragma experimental ABIEncoderV2;
+		contract C {
+			function f() public pure returns (bytes memory, bytes memory) {
+				return (abi.encode(1, -2), abi.encodePacked(int8(1), int8(-2)));
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp]: I guess we could remove the `abi.encodePacked` part because it doesn't test the case.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Typo.

**Comment by axic** [comment]
@chriseth sorry, didn't see your review, but this should be mergeable now.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -12527,6 +12527,42 @@ BOOST_AUTO_TEST_CASE(abi_encode_empty_string_v2)
...
+}
+
+BOOST_AUTO_TEST_CASE(abi_encode_rational_v2)
+{
+	char const* sourceCode = R"(
+		// Tests that rational numbers (even negative ones) are encoded properly.
+		pragma experimental ABIEncoderV2;
+		contract C {
+			function f() public pure returns (bytes memory) {
+				return (abi.encode(1, -2);
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: still typo

Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Damn.

**Review by chriseth — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**